pkg/services/ngalert/notifier/channels/wecom.go

```diff
@@ -5,40 +5,82 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"net/http"
+	"time"
 
 	"github.com/prometheus/alertmanager/template"
 	"github.com/prometheus/alertmanager/types"
+	"golang.org/x/net/context/ctxhttp"
+	"golang.org/x/sync/singleflight"
 
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/notifications"
 )
 
+var WeComEndpoint = "https://qyapi.weixin.qq.com"
+
+const defaultWeComChannelType = "groupRobot"
+const defaultWeComMsgType = "markdown"
+const defaultWeComToUser = "@all"
+
 type wecomSettings struct {
-	URL     string `json:"url" yaml:"url"`
-	Message string `json:"message,omitempty" yaml:"message,omitempty"`
-	Title   string `json:"title,omitempty" yaml:"title,omitempty"`
+	channel     string
+	EndpointURL string `json:"endpointUrl,omitempty" yaml:"endpointUrl,omitempty"`
+	URL         string `json:"url" yaml:"url"`
+	AgentID     string `json:"agent_id,omitempty" yaml:"agent_id,omitempty"`
+	CorpID      string `json:"corp_id,omitempty" yaml:"corp_id,omitempty"`
+	Secret      string `json:"secret,omitempty" yaml:"secret,omitempty"`
+	MsgType     string `json:"msgtype,omitempty" yaml:"msgtype,omitempty"`
+	Message     string `json:"message,omitempty" yaml:"message,omitempty"`
+	Title       string `json:"title,omitempty" yaml:"title,omitempty"`
+	ToUser      string `json:"touser,omitempty" yaml:"touser,omitempty"`
 }
 
 func buildWecomSettings(factoryConfig FactoryConfig) (wecomSettings, error) {
-	var settings = wecomSettings{}
+	var settings = wecomSettings{
+		channel: defaultWeComChannelType,
+	}
 
 	err := factoryConfig.Config.unmarshalSettings(&settings)
 	if err != nil {
 		return settings, fmt.Errorf("failed to unmarshal settings: %w", err)
 	}
 
-	if settings.Message == "" {
+	if len(settings.EndpointURL) == 0 {
+		settings.EndpointURL = WeComEndpoint
+	}
+
+	if len(settings.MsgType) == 0 {
+		settings.MsgType = defaultWeComMsgType
+	}
```

Comment:
We can validate the enum here to protect against buggy configs. I wish we had a logger in these methods so we could log the config problem if we write down "maredown" or something else incorrect.
```go
if !settings.MsgType.IsValid() {
   settings.MsgType = defaultWeComMsgType
}
```